ddtrace/contrib/aiohttp/patch.py

```diff
@@ -1,40 +1,375 @@
+            resp = yield from self._handle_response(  # noqa: E999
+                self.__wrapped__.__aenter__())
+            return resp
+
+        @asyncio.coroutine
+        def __aexit__(self, exc_type, exc_val, exc_tb):
+            try:
+                resp = yield from self.__wrapped__.__aexit__(exc_type, exc_val,  # noqa: E999
+                                                             exc_tb)
+            finally:
+                if self._self_have_context and self._self_trace_context:
+                    self._self_span.finish()
+            return resp
+
+
+def _create_wrapped_request(method, enable_distributed, trace_headers,
+                            trace_context, func, instance, args, kwargs):
+    # Use any attached tracer if available, otherwise use the global tracer
+    pin = Pin.get_from(instance)
+
+    # bail on the tracing if not enabled.
+    if not pin.tracer.enabled:
+        return func(*args, **kwargs)
+
+    if method == 'REQUEST':
+        if 'method' in kwargs:
+            method = kwargs['method']
+        else:
+            method = args[0]
+
+        if 'url' in kwargs:
+            url = URL(kwargs['url'])
+        else:
+            url = URL(args[1])
+    else:
+        if 'url' in kwargs:
+            url = URL(kwargs['url'])
+        else:
+            url = URL(args[0])
+
+    if should_skip_request(pin, url):
+        result = func(*args, **kwargs)
+        return result
+
+    # Create a new span and attach to this instance (so we can
+    # retrieve/update/close later on the response)
+    # Note that we aren't tracing redirects
+    span = pin.tracer.trace(
+        'ClientSession.request', service=_get_service_fallback(pin),
+        span_type=ext_http.TYPE)
+
+    if enable_distributed:
+        headers = kwargs.get('headers', {})
+        if headers is None:
+            headers = {}
+        propagator = HTTPPropagator()
+        propagator.inject(span.context, headers)
+        kwargs['headers'] = headers
+
+    _set_request_tags(span, url)
+    span.set_tag(ext_http.METHOD, method)
+
+    obj = _WrappedRequestContext(
+        func(*args, **kwargs), pin, span, trace_headers, trace_context)
+    return obj
+
+
+def _create_wrapped_response(client_session, trace_headers, cls, instance,
+                             args, kwargs):
+    obj = _WrappedResponseClass(cls(*args, **kwargs),
+                                Pin.get_from(client_session), trace_headers)
+    return obj
+
+
+def _wrap_clientsession_init(trace_headers, func, instance, args, kwargs):
+    # Use any attached tracer if available, otherwise use the global tracer
+    pin = Pin.get_from(instance)
+
+    # bail on the tracing if not enabled.
+    if not pin.tracer.enabled:
+        return func(*args, **kwargs)
+
+    response_class = kwargs.get('response_class', aiohttp.ClientResponse)
+
+    loop = kwargs.get('loop')
+    if not loop:
+        loop = asyncio.get_event_loop()
+
+    connector = kwargs.get('connector')
```

Comment:
We should also check if it's passed through args

Reply:
btw clientsession doesn't have args, just checked, only kwargs